### ISSUE TITLE
Book class could use a \part command

### ISSUE THREAD

**Issue by alerque** [simoncozens/sile#601]
Not all books use `\chapter` as their top level division, and having a `\part` command that behaved as expected would ease transition from LaTeX. :stuck_out_tongue_winking_eye:

**Comment by simoncozens** [comment]
I'm going to need this in rewriting the documentation; the new docs will be in two parts, a users' guide and a developers' guide.

**Comment by alerque** [comment]
I have one in CaSILE. I'll work on porting the rest of my possibly useful book class changes some next week.

Question: What comes above "part"? I have an current book project where the author (nameless to protect the guilty) has insisted they have two levels of groupings above "chapter". I've been typesetting chapters as sections to compensate, but I've basically had to copy all the functions and logic down the tree and it seems like an accommodation for "collection" works that contain other whole books (including ones with parts) would be an okay thing to have out of the box.

**Comment by simoncozens** [comment]
volume?

**Comment by alerque** [comment]
# This is a rabbit hole

Adding a `\part` command is quite logical. As long as I was messing with the book class I figured I should expand it with some more levels. The result is going to require a little more tweaking than just adding commands.

Adding `\part` bumps the numbering level of `\chapter` which will be a breaking change. We can paint over most of it by changing the default numbering system to not number parts, but I'm not sure that's actually desired. We probably need some command to set what the top level numbered division is.

As long as we are fixing the logic to handle `\part` I figure adding `\volume` is a good idea. Now I'm trying to figure out what to call lower levels. This is somewhat important because I'm working on the Pandoc reader/writer and need to pick what levels map to what by default.

## Markdown doesn't care

* Almost all implementations support at least 6 levels
* Many support an arbitrary number of levels

## HTML supports 6.5 levels

* `<h1>` through `<h6>`
* A one off arbitrary `<title>` tag

## LaTeX is a monster

* `\part` (typically level -1)
* `\chapter` (typically level 0)
* `\section` (typically level 1)
* `\subsection`
* `\paragraph`
* `\subparagraph`

## Word Processors conflate rhyme with reason

* Google Docs & Libre Office both have 3 levels of headings, plus `Title` and `Subtitle` for good measure, for a total of 4.5?
* Microsoft Word ... I don't have this
* Apple Pages ... I don't have this either

**Comment by simoncozens** [comment]
Why do we need to renumber the level of chapter? What's the problem with having levels 0 and -1? That's the nice thing about real integers - they can extend as far as you like in either direction.

**Comment by alerque** [comment]
Good question. The counter formatting code will still need some better clues but we can probably do that. I'll review...

**Comment by alerque** [comment]
Not sure if there is anything useful, but it looks like I started on this in [this old branch](https://github.com/alerque/sile/tree/texishness) that was never made into a PR. Not vouching for the code, just something to check when working on this.

**Comment by Omikhleia** [comment]
In the "LaTeX is a monster" rabbit hole, and also as kind of an answer to a question above regarding what comes above a "part"...

LaTeX also has `\frontmatter` (for preamble, defaults to pages in roman i, ii, iii, iv, etc., numbered chapters), `mainmatter` (arabic page numbering, reset to 1, normal chapter numbering, etc.), `\appendix` (chapter numbering reset to A, B, C, etc.) and `backmatter` (for end sections before e.g. bibliography and indexes) :smile_cat: 

All of these can contain the other above-mentioned sectioning commands, with different effects, if I remember correctly.

In my (pet) projects, I would need all the sub-sectioning to go that deep from front/main/back matter down to sub-sub-sections (lol), but aside from say parts and chapters, higher-level sections (such as front matter, etc.) would make sense, with a somewhat different page template (different placing of frames, different running headers, etc.)

**Comment by simoncozens** [comment]
Thought: if the presenting issue is “the book class needs to support lots of different ways to do section numbering”, perhaps the underlying issue is “it is still too difficult for people to generate their own custom book class.”

**Comment by Omikhleia** [comment]
> Thought: if the presenting issue is “the book class needs to support lots of different ways to do section numbering”, perhaps the underlying issue is “it is still too difficult for people to generate their own custom book class.”

Or people expect a full-fledge book class from the start :) But on one hand: TeX 1978 had nothing like a "book" class in the usual sense. LaTeX 1984 got a book class... though front/back matters etc. only made it with LaTeX2e in 1994. Also noting weird "extra possibilities for section headings" in 1996, etc. the list could go on.

Ok comparing apples and oranges in a way, but it's still: hard to do well, time-consuming and likely not fun for _most_ people (who would want to typeset, not to dabble into the internals of class and packages). Whether it is really difficult with SILE, I'll give it a try...

**Comment by alerque** [comment]
Yes the class system needs improvements. The inheritance model will improve greatly when I finish eradicating `std` from the mix. Sorry that has gotten pushed back a few times (see #1065). The work isn't that hard to do but doing it while providing an upgrade path for anybody that had used the old inheritance model is hard.

In the mean time CaSILE's book class `cabook` has had `\part` forever and is used at industrial scale. Frankly I'm 100% convinced this should be part of the default `book` class. It's a nearly ubiquitous concept in publishing and used by many other document formats. My Pandoc writer for example assumes it because it many other document formats do. It turns out these days Markdown is one of the major ones that does not.

**Comment by Omikhleia** [comment]
See discussion #2330
